### PR TITLE
Use yaml config to manage k8s resources deployed by bootstrapper

### PR DESCRIPTION
**Pull request by kunmingg**
This provide us flexibility to control kubeflow resources when facing different code version / env. 
solve https://github.com/kubeflow/kubeflow/issues/829

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/853)

### CODE REVIEW

File: bootstrap/cmd/bootstrap/app/options/options.go [hunk 1]
```diff
@@ -26,11 +26,9 @@ type ServerOption struct {
 	InCluster     bool
 	KeepAlive     bool
 	AppDir        string
-	KfVersion     string
-	NameSpace     string
-	Project       string
+	ConfigDir	  string
```
Comment (jlewi) [inline bootstrap/cmd/bootstrap/app/options/options.go]: Why a directory and not a single file?

File: bootstrap/cmd/bootstrap/app/server.go [hunk 2]
```diff
@@ -205,7 +249,7 @@ func setupNamespace(namespaces type_v1.NamespaceInterface, name_space string) er
 
 func createComponent(opt *options.ServerOption, kfApp *kApp.App, fs *afero.Fs, args []string) {
 	componentName := args[1]
-	componentPath := filepath.Join(opt.AppDir, "components", componentName+".jsonnet")
+	componentPath := filepath.Join(opt.AppDir, "Components", componentName+".jsonnet")
```
Comment (jlewi) [inline bootstrap/cmd/bootstrap/app/server.go]: Why is Components capitalized?

File: bootstrap/cmd/bootstrap/app/server.go [hunk 3]
```diff
@@ -221,6 +265,58 @@ func createComponent(opt *options.ServerOption, kfApp *kApp.App, fs *afero.Fs, a
...
 
+func appGenerate(opt *options.ServerOption, kfApp *kApp.App, fs *afero.Fs, appConfig *AppConfig) {
+	libs, err := (*kfApp).Libraries()
+
+	if err != nil {
+		log.Fatalf("Could not list libraries for app; error %v", err)
+	}
+	// Install packages.
+	for _, p := range appConfig.Packages {
+		//for _, p := range []string{"kubeflow/core", "kubeflow/tf-serving", "kubeflow/tf-job", "kubeflow/pytorch-job"} {
```
Comment (jlewi) [inline bootstrap/cmd/bootstrap/app/server.go]: Delete commented out code.

File: bootstrap/config/gcp_prototype.yaml [hunk 4]
```diff
@@ -0,0 +1,44 @@
+# Sample config for kubeflow bootstrapper
+---
+# Apps that always apply
+DefaultApp:
```
Comment (jlewi) [inline bootstrap/config/gcp_prototype.yaml]: Why do the key names start with capital letters? This doesn't follow standard K8s or GCP deployment manager syntax. Can we make it lower case?

File: bootstrap/config/gcp_prototype.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
+# Sample config for kubeflow bootstrapper
+---
+# Apps that always apply
```
Comment (jlewi) [inline bootstrap/config/gcp_prototype.yaml]: Why are we creating multiple apps? Why wouldn't we just create a single app and use the ConfigMap to specify the app we want to create?

File: bootstrap/cmd/bootstrap/app/server.go [hunk 6]
```diff
@@ -45,19 +45,63 @@ import (
...
+}
+
+//
+type AppConfig struct {
+	Packages   []KsPackage
+	Components []KsComponent
+	Parameters []KsParameter
+}
+
+type BootConfig struct {
```
Comment (jlewi) [inline bootstrap/cmd/bootstrap/app/server.go]: Why are there multiple apps? And if you want to support multiple apps why wouldn't you use a container like a list or map.

My sense though is we should start with a single app unless we have a clear use case for multiple ones.

I think if a user wanted multiple apps they could run bootstrapper multiple times

File: bootstrap/cmd/bootstrap/app/server.go (hunk 2 above)
Comment (kunmingg) [reply bootstrap/cmd/bootstrap/app/server.go]: goland's auto rename...

**Comment by jlewi** [comment]
Awesome; thanks for the quick fix.

/lgtm
/approve

**Pull request merged**